src/h_vialib/exceptions.py

```diff
@@ -0,0 +1,13 @@
+"""Exceptions."""
+
+
+class InvalidToken(Exception):
+    """A security check has failed on a token."""
+
+
+class MissingToken(InvalidToken):
+    """An expected token is missing.
+
+    The missing token is a child of `InvalidToken` so you can distinguish
+    between the two, or catch them together as you see fit
+    """
```

Comment:
I think this makes it easy to write nonsensical code like:

```python
try:
    ...
except MissingToken:
    ...
except InvalidToken:
    ...
```

The code is trying to handle missing and invalid tokens differently but the `except InvalidToken` will never be hit because `MissingToken` is an `InvalidToken` subclass. If the two `except`'s were in the other order it'd work. I'm not sure whether PyLint or Python itself would catch this mistake.

Having a separate `TokenException` base class that both `InvalidToken` and `MissingToken` inherit from would mean that you can catch `InvalidToken` and `MissingToken` separately and it doesn't matter what order you put the `except`'s in, and if you _want_ to catch both with a single `except` you'd catch the `TokenException` base class instead.

Our standard guide for laying out exception classes would achieve this "automatically" (without even having to think of the `TokenException` base class) because all `vialib` exceptions would inherit from `ViaLibException` and all `vialib.secure` exceptions would inherit from `SecureException`. So you could catch `InvalidToken` and `MissingToken` separately in either order, or catch either `SecureException` or `ViaLibException` if you want to catch both. It'd also put `InvalidToken` and `MissingToken` closer to where they're used, in `secure/exceptions.py`. The separate `TokenException` _can_ still be added but isn't necessary.

```python
# h_vialib/exceptions.py

class ViaLibException(Exception):
    """Base class for all h_vialid exceptions."""
```

```python
# h_vialib/secure/exceptions.py

from h_vialib.exceptions import ViaLibException

class SecureException(ViaLibException):
    """Base class for h_vialib.secure exceptions."""

# You wouldn't really need this one anymore, but you could still do it.
class TokenException(SecureException):
    """Base class for secure token-related exceptions."""

class InvalidToken(SecureException):
    """A security check has failed on a token."""

class MissingToken(InvalidToken):
    """An expected token is missing."""
```

Reply:
The minimal change I think would be:

```python
# h_vialib/exceptions.py

class ViaLibException(Exception):
    """Base class for h_vialib exceptions."""

class InvalidToken(ViaLibException):
    """A security check has failed on a token."""

class MissingToken(ViaLibException):
    """An expected token is missing."""
```

Rather than the full version with two separate `exceptions.py` files. The `TokenException` base class in the middle would still be optional

Reply:
I don't think that's how exception handling works, otherwise every catch statement would trigger on raising `BaseException`. A `MissingToken` is an instance of `InvalidToken`, but an `InvalidToken` is not an instance of `MissingToken`.

The following code could do something like what you are suggesting:

```
try:
    ...
except InvalidToken:
    ...
except MissingToken:
    ...
```
But this is true of any inheritance in exceptions, including the fix suggested here:

```
try:
    ...
except TokenException:
    ...
except MissingToken:
    ...
```

Reply:
Yes, with a separate base class the order of exceptions still matters, but I'm saying that the separate base class makes the mistake easier to spot / harder to make. You always have to look at a list of `except`'s and make sure that they're in the right order: base classes must come last. With a separate base class like `TokenException` it's more obvious just from the name that it's a base class. Whereas with `MissingToken` and `InvalidToken` it's surprising that `InvalidToken` is actually a base class and must come last. You wouldn't spot it.

To put it another way: there's no reason to catch a `TokenException` base class _unless_ you want a catch-all `except` for all token-related errors

Reply:
I think the above is just a boo boo. @seanh explained in slack he does mean the case I'm suggesting here. Basically if the exceptions don't have an implied hierarchy in the name it makes it easier to accidentally catch calling a parent above a child.

I'm ok with introducing hierarchy for exceptions you want to catch (like this token one), but I don't think there's a reason to introduce a `ViaLibException`. Exceptions that come from `vialib` have no particular semantic meaning in the general case (unlike `requests` or something for example).

Reply:
Here's the reason why this should have a `VialibException` base class:

By adding a `TokenException` base class and having `InvalidToken` and `MissingToken` inherit from it you've captured all the advantages of our standard guide for exception classes (the `VialibException` approach) _in this particular case_ but to do that you had to spot the relation between `MissingToken` and `InvalidToken`, realise that they should be in an inheritance hierarchy and that there should be a third `TokenException` base class. There are lots of ways to get it wrong: `MissingToken` inheriting from `InvalidToken`, or neither of them inheriting from anything shared parent (except `Exception`), inheriting from Pyramid's exception classes, etc.

If we follow a standard guide that every package has a base `FoopackageException` that all its exceptions inherit from then, despite the semantic meaninglessness of `FoopackageException`, you get all the practical benefits without all the thinking and opportunities for creative ways to mess it up:

* You can catch `InvalidToken` and `MissingToken` separately, in any order
* If you want to catch both, you can catch `FoopackageException`
* If you come along later and add a new `ExpiredToken` class my code that catches `FoopackageException` will still work
* You can still add the intermediate `TokenException` base class if you want, but it's no longer necessary
* The guide also encourages exception classes to be defined not too near to and not too far from the code that raises them, to have not too many and not too few separate exception classes

Basically I think if everyone just follows the guide we'll be alright, whereas if everyone solves this problem anew for themselves we'll end up with an exciting variety of problems